docs/2.9.0/docs/query/pqs-user-guide.rst

```diff
@@ -1,466 +1,1741 @@
+   PostgreSQL's MVCC provides consistency benefits without the user
+   needing to understand the underlying implementation, we seek to
+   provide a similar experience for readers of the ledger data.
+-  **Idempotent**: PQS is designed to be restarted at any time, and will
+   not impact the integrity of the data. This is achieved by using
+   offsets to ensure clean restarts, and ensuring that (other than the
+   database itself) PQS is stateless.
+-  **Watermarks**: a single thread maintains a watermark denoting the
+   most recent contigous transaction - representing the offset of the
+   ledger that is consistent and available. In addition, the watermark
+   processes the "archive" mutation on any archived contracts or
+   interface views, on a batched basis. This reintroduces data
+   consistency without needing readers to perform pathalogical table
+   scans to achieve this. This efficiently resolves the uncertainty
+   created by the parallel writes.
+-  **Schemaless content**: content defined in Daml templates uses the
+   JSONB datatype to store the data. This allows for a schemaless
+   approach, and is able to store any Daml model without needing to
+   change the schema - other than custom JSONB indexes.
+
+Resilience
 ==========
 
-The PQS is intended for continuous operation. Upon restart after an interruption, PQS determines the last consistent offset and continues incremental processing from that point onward. PQS terminates when encountering any error and leaves it up to the orchestration layer (such as Kubernetes) or the operator to determine the appropriate course of action.
+PQS operates as a long-running process, and faces many potential threats
+to its availability. We would expect PQS on occasion to encounter issues
+with dependencies during operation. Three mechanisms can assist in
+ensuring resilience to such failures:
+
+-  **Redundancy** involves running multiple instances of PQS in parallel
+   to ensure that the system remains available even if one instance
+   fails.
+-  **Retry** involves healing from transient and recoverable failures
+   without shutting down the process or requiring operator intervention.
+-  **Recovery** entails reconciling the current state of the ledger with
+   already exported data in the datastore after a cold start, and
+   continuing from the latest checkpoint.
 
 High Availability
-=================
+-----------------
+
+Multiple isolated instances of PQS can be instantiated without any
+cross-dependency. This allows for an active-active high availability
+clustering model. Please note that different instances might not be at
+the same offset due to different processing rates and general network
+non-determinism. PQS's Read API provides capabilities to deal with this
+'eventual consistency' model, to ensure that readers have at least
+'repeatable read' consistency. See ``validate_offset_exists()`` in the
+`Read API <004-read-api.md#offset-management>`__ for more details.
+
+|image3|
```

Comment:
could this figure have "gRPC" and "JDBC" labelled so that it is easy to visualize?  Readers hate to think.